sonar-dotnet-shared-library/src/main/java/org/sonarsource/dotnet/shared/plugins/FileTypeSensor.java

```diff
@@ -0,0 +1,95 @@
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarsource.dotnet.shared.plugins;
+
+import java.util.Optional;
+import org.sonar.api.batch.ScannerSide;
+import org.sonar.api.batch.fs.FileSystem;
+import org.sonar.api.batch.fs.InputFile.Type;
+import org.sonar.api.batch.sensor.Sensor;
+import org.sonar.api.batch.sensor.SensorContext;
+import org.sonar.api.batch.sensor.SensorDescriptor;
+import org.sonar.api.config.Configuration;
+import org.sonar.api.utils.log.Logger;
+import org.sonar.api.utils.log.Loggers;
+
+/**
+ * This class is a non-global sensor used to count the type of files in the .NET projects (i.e. Scanner modules).
+ *
+ * Why is this needed?
+ * - the Scanner for MSBuild categorizes projects as MAIN or TEST (see https://github.com/SonarSource/sonar-scanner-msbuild/wiki/Analysis-of-product-projects-vs.-test-projects)
+ * - in SQ / SC, users can specify which files should be considered as MAIN (sources) or TEST (test sources) (see https://docs.sonarqube.org/latest/project-administration/narrowing-the-focus/)
+ * - the categorization is not obvious, so this additional information should help users debug when needed
+ */
+@ScannerSide
+public class FileTypeSensor implements Sensor {
+  private static final Logger LOG = Loggers.get(FileTypeSensor.class);
+
+  private final ProjectTypeCollector projectTypeCollector;
+  private final DotNetPluginMetadata pluginMetadata;
+
+  public FileTypeSensor(ProjectTypeCollector projectTypeCollector, DotNetPluginMetadata pluginMetadata) {
+    this.projectTypeCollector = projectTypeCollector;
+    this.pluginMetadata = pluginMetadata;
+  }
+
+  @Override
+  public void describe(SensorDescriptor descriptor) {
+    String name = String.format("Verify what types of files (MAIN, TEST) are in %s projects.", pluginMetadata.languageKey());
+    descriptor.name(name);
+    // we do not filter by language because we want to be called on projects without sources
+    // (that could reference only shared sources e.g. in .projitems)
+  }
+
+  @Override
+  public void execute(SensorContext context) {
+    FileSystem fs = context.fileSystem();
+    Configuration configuration = context.config();
+
+    boolean hasMainFiles = hasFilesOfType(fs, Type.MAIN);
+    boolean hasTestFiles = hasFilesOfType(fs, Type.TEST);
+
+    Optional<String> projectName = configuration.get("sonar.projectName");
+    if (projectName.isPresent()) {
+      LOG.debug("Adding file type information (has MAIN '{}', has TEST '{}') for project '{}' (base dir '{}'). For debug info, see ProjectInfo.xml in '{}'.",
+        hasMainFiles, hasTestFiles, projectName.get(), getValueOrEmpty(configuration.get("sonar.projectBaseDir")), getValueOrEmpty(getProjectOutPath(configuration)));
+      projectTypeCollector.addProjectInfo(hasMainFiles, hasTestFiles);
+    }
+  }
+
+  // this is duplicated from future DotNetSensor
+  private boolean hasFilesOfType(FileSystem fs, Type fileType) {
```

Comment:
ok, I missed this after rebase. I will refactor to avoid duplication after I get the review comments, together